tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/CopyReferenceOnlyPackages.cs

```diff
@@ -0,0 +1,132 @@
+        /// 
+        /// {PackageCacheDir}/{lowercase id}/{version}/{lowercase id}.{version}.nupkg
+        ///
+        /// This code assumes that these are expanded packages loaded from nuget.
+        /// </summary>
+        [Required]
+        public string PackageCacheDir { get; set; }
+
+        /// <summary>
+        /// The destination directory for the dlls in reference packages.
+        /// Paths are preserved in the destination.
+        /// </summary>
+        [Required]
+        public string DllDestinationDir { get; set; }
+
+        /// <summary>
+        /// The destination directory for the nupkg files identified 
+        /// as containing reference-only packages.
+        /// </summary>
+        [Required]
+        public string IdentifiedPackagesDir { get; set; }
+
+        /// <summary>
+        /// The destination directory for the reference packages.
+        /// </summary>
+        [Required]
+        public string DestinationDir { get; set; }
+
+        /// <summary>
+        /// Enumerate all files in a directory and its sub-directories.
+        /// </summary>
+        private static IEnumerable<string> EnumerateAllFiles(string path, string searchPattern)
+        {
+            return Directory.EnumerateFiles(path, searchPattern, SearchOption.AllDirectories);
+        }
+
+        public override bool Execute()
+        {
+            DateTime startTime = DateTime.Now;
+
+            var referenceOnlyPackageDirectories = EnumerateAllFiles(PackageCacheDir, "*.nupkg")
+                .Where(nupkgFilePath =>
+                {
+                    // Get all files in the nupkg path and normalize directory separators
+                    var nupkgFiles = EnumerateAllFiles(Path.GetDirectoryName(nupkgFilePath), "*.*").ToArray();
+
+                    // Do not include directories that contain exes, shared object files, OSX dynamic libraries
+                    // or profiling data
+                    if (nupkgFiles
+                        .Any(
+                            file => extensionsToExclude.Contains(Path.GetExtension(file)) 
+                            || pathsToExclude.Any(path => file.Contains(path))))
+                    {
+                        return false;
+                    }
+                    
+                    // Return directories that, if containing dlls, only have dlls in the
+                    // ref folder
+                    return nupkgFiles
+                            .Where(file => String.Equals(Path.GetExtension(file), ".dll", StringComparison.OrdinalIgnoreCase))
+                            .All(dir => dir.Contains(refPath));
+                })
+                .Select(Path.GetDirectoryName)
+                .ToArray();
+
+            Directory.CreateDirectory(IdentifiedPackagesDir);
+            foreach (var dir in referenceOnlyPackageDirectories)
+            {
+                foreach (var file in EnumerateAllFiles(dir, "*.*"))
```

Comment:
In the earlier iteration I suggested keeping the enumeration results from `nupkgFiles` rather than repeating it here potentially using an anonymous type--understandable if it's not worth it and that's why it's marked resolved, but pointing out it's still here.

Reply:
I think I'll leave this.  I'm not sure it's worth changing things around at this point.